notebooks/03-query/09-functions-case.ipynb

```diff
@@ -0,0 +1,499 @@
+    "    CASE\n",
+    "        WHEN \"event_subtype\" <= 24 THEN '1'\n",
+    "        WHEN \"event_subtype\" BETWEEN 25 AND 49 THEN '2'\n",
+    "        WHEN \"event_subtype\" BETWEEN 50 AND 74 THEN '3'\n",
+    "        WHEN \"event_subtype\" BETWEEN 75 AND 99 THEN '4'\n",
+    "        ELSE NULL\n",
+    "        END\n",
+    "    END\n",
+    "    AS \"funnelStage\",\n",
+    "  CASE WHEN (\"event_type\" = 'GoodLoad') THEN 1 ELSE 0 END AS \"isImpression\",\n",
+    "  CASE WHEN (\"event_subtype\" IS NULL AND \"event_type\" = 'LayerClear') THEN 1 ELSE 0 END AS \"isConversion\",\n",
+    "  COUNT(*) AS \"events\"\n",
+    "FROM \"ext\"\n",
+    "GROUP BY 1, 2, 3, 4, 5, 6, 7, 8, 9\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "\n",
+    "display.run_task(sql)\n",
+    "sql_client.wait_until_ready('example-koalas-conditions-rollup')\n",
+    "display.table('example-koalas-conditions-rollup')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c22100c0-b453-44bc-ad0d-76b2e6a3f261",
+   "metadata": {},
+   "source": [
+    "A query like the following then gives us a good indicator of people's journeys."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "dc788b8a-f6c4-4f7e-a4cf-1444ace4303e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT\n",
+    "  TIME_FLOOR(__time, 'PT4H') AS \"timebucket\",\n",
+    "  SUM(\"isImpression\") AS \"totalImpressions\",\n",
+    "  COUNT(DISTINCT \"session\") FILTER (WHERE \"funnelStage\" = 1) AS \"reached_stage1\",\n",
+    "  COUNT(DISTINCT \"session\") FILTER (WHERE \"funnelStage\" = 2) AS \"reached_stage2\",\n",
+    "  COUNT(DISTINCT \"session\") FILTER (WHERE \"funnelStage\" = 3) AS \"reached_stage3\",\n",
+    "  COUNT(DISTINCT \"session\") FILTER (WHERE \"funnelStage\" = 4) AS \"reached_stage4\",\n",
+    "  SUM(\"isConversion\") AS \"totalConversions\"\n",
+    "FROM \"example-koalas-conditions-rollup\"\n",
+    "WHERE TIME_IN_INTERVAL(__time,'2019-08-25T0/P1D')\n",
+    "GROUP BY 1\n",
+    "ORDER BY 1 ASC\n",
+    "'''\n",
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "44738d6d-cec2-40ad-aaba-998c758c63f4",
+   "metadata": {},
+   "source": [
+    "## Clean up\n",
+    "\n",
+    "Run the following cell to remove the XXX used in this notebook from the database."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8082b545-ba7f-4ede-bb6e-2a6dd62ba0d8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "druid.datasources.drop(\"example-koalas-conditions\")\n",
+    "druid.datasources.drop(\"example-koalas-conditions-rollup\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "54b8d5fe-ba85-4b5b-9669-0dd47dfbccd1",
+   "metadata": {},
+   "source": [
+    "## Summary\n",
+    "\n",
+    "* There are two forms of CASE, simple and search\n",
+    "* CASE statements can be used to transform query results and incoming data\n",
```

Comment:
Did you want to add a summary point on more of the fancy applications of CASE? The last 1/3 or so of the notebook seemed to focus on this and a high level refresher would be useful.